### PR TITLE
Retrieve upstream tarballs from GitHub release assets

### PR DESCRIPTION
**Pull request by mkoeppe**
We propose to add GitHub release assets as a source from which upstream tarballs can be retrieved. This would take priority over trying the Sage mirrors, thus reducing the reliance of the project on self-maintained file servers. https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases#storage-and-bandwidth-quotas

For the purpose of testing this PR, one tarball missing on the mirrors (https://github.com/sagemath/sage/pull/36381#issuecomment-1752268975) has been uploaded to https://github.com/sagemath/sage/releases/tag/10.1

Creating GitHub releases and depositing the upstream tarballs can later be automated in a GH Actions workflow, e.g., using https://github.com/softprops/action-gh-release. As we don't make GitHub releases for beta versions, the previous stable release can be used.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: build/sage_bootstrap/download/mirror_list.py [hunk 1]
```diff
@@ -41,17 +41,66 @@ class MirrorListException(RuntimeError):
...
+                    for line in f:
+                        line = line.replace('${SAGE_ROOT}', SAGE_ROOT)
+                        line = line.replace('${SAGE_DISTFILES}', SAGE_DISTFILES)
+                        if '${SAGE_SERVER}' in line:
+                            SAGE_SERVER = os.environ.get("SAGE_SERVER", "")
+                            if not SAGE_SERVER:
+                                continue
+                            line = line.replace('${SAGE_SERVER}',)
+                        line = line.strip()
+                        if line.startswith('#'):
```
Comment (jhpalmieri) [inline build/sage_bootstrap/download/mirror_list.py]: Speed probably doesn't matter too much, but wouldn't it be a little faster to move `if not line` and `if line.startswith('#')` earlier in this block?

File: src/bin/sage-update-version [hunk 2]
```diff
@@ -87,6 +87,25 @@ EOF
...
+) | uniq | head -n 3 > "$SAGE_ROOT/.upstream.d/20-github.com-sagemath-sage-releases.tmp"
+( cat <<EOF
+# Upstream packages as uploaded as GitHub release assets.
+# This file is automatically updated by the sage-update-version script.
+EOF
+  cat "$SAGE_ROOT/.upstream.d/20-github.com-sagemath-sage-releases.tmp"
+) > "$SAGE_ROOT/.upstream.d/20-github.com-sagemath-sage-releases"
+rm -f "$SAGE_ROOT/.upstream.d/20-github.com-sagemath-sage-releases.tmp"
+
+exit
```
Comment (jhpalmieri) [inline src/bin/sage-update-version]: Why `exit`? Doesn't that end the script before executing the rest of it? Also, do we need to `git commit` the new file?

Comment (mkoeppe) [reply src/bin/sage-update-version]: Whoops, that was debugging code. Fixed in 371be50

File: build/sage_bootstrap/download/mirror_list.py (hunk 1 above)
Comment (mkoeppe) [reply build/sage_bootstrap/download/mirror_list.py]: Thanks, done in 7e65260

**Comment by jhpalmieri** [comment]
In principle it looks okay to me. I'm guessing that github will be a reliable place to store tarballs. Are there any downsides?

**Comment by mkoeppe** [comment]
I'm not aware of any downsides of this

**Comment by jhpalmieri** [comment]
Looks good, let's merge it.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by mkoeppe** [comment]
I'm marking it as a blocker [so that it takes effect in current CI runs](https://github.com/sagemath/sage/wiki/Sage-10.2-Release-Tour#open-blocker-prs-are-applied-automatically-in-ci-workflows), which currently can get stuck because of the missing tarballs (see https://github.com/sagemath/sage/actions/runs/6454925852/job/17521487027#step:8:14136)

**Comment by dimpase** [comment]
why is 10.1 hardcoded in [.upstream.d/20-github.com-sagemath-sage-releases](https://github.com/sagemath/sage/pull/36435/files#diff-5a062b99560e27ee413a97cafabf307e02398acb9afaff322714df3ac5e441a8) ?

**Review by dimpase — APPROVED** [review]
OK, good

**Pull request merged**